File: src/main/java/scrabble/phrases/words/Adjective.java
package scrabble.phrases.words;

import java.util.Objects;

/**
 * Romanian adjective with feminine form derivation.
 *
 * @param word the adjective in masculine form
 * @param syllables the number of syllables
 * @param rhyme the rhyme suffix (last 3 characters)
 */
public record Adjective(
    String word,
    int syllables,
    String rhyme
) implements Word {

    /**
     * Creates a new Adjective with computed syllables and rhyme.
     *
     * @param word the adjective in masculine form
     */
    public Adjective(String word) {
        this(
            word,
            WordUtils.computeSyllableNumber(word),
            WordUtils.computeRhyme(word)
        );
    }

    /**
     * Compact constructor for validation.
     */
    public Adjective {
        Objects.requireNonNull(word, "word cannot be null");
    }

    /**
     * Gets the feminine form of the adjective.
     *
     * @return the feminine form
     */
    public String feminine() {
        if (word.endsWith("esc")) {
            return word.substring(0, word.length() - 2) + "ască";
        }
        if (word.endsWith("eț")) {
            return word.substring(0, word.length() - 1) + "ață";
        }
        if (word.endsWith("or")) {
            return word.substring(0, word.length() - 1) + "are";
        }
        if (word.endsWith("os")) {
            return word.substring(0, word.length() - 1) + "asă";
        }
        if (word.endsWith("iu")) {
            return word.substring(0, word.length() - 1) + "e";
        }
        if (word.endsWith("ci")) {
            return word.substring(0, word.length() - 1) + "e";
        }
        if (word.endsWith("ru")) {
            return word.substring(0, word.length() - 1) + "ă";
        }
        if (word.endsWith("e") || word.endsWith("o") || word.endsWith("i")) {
            return word;
        }
        return word + "ă";
    }

    // Legacy method names for compatibility
    public String getWord() {
        return word;
    }

    public int getSyllables() {
        return syllables;
    }

    public String getRhyme() {
        return rhyme;
    }

    public String getFeminine() {
        return feminine();
    }
}

File: src/main/kotlin/scrabble/phrases/words/Adjective.kt
package scrabble.phrases.words

data class Adjective(
    override val word: String,
    override val syllables: Int = WordUtils.computeSyllableNumber(word),
    override val rhyme: String = WordUtils.computeRhyme(word),
    val feminine: String = computeFeminine(word),
    val feminineSyllables: Int = WordUtils.computeSyllableNumber(feminine)
) : Word {

    /** Returns the appropriate adjective form for a noun of given gender.
     *  Romanian neuter singular uses the feminine adjectival form; masculine uses base word.
     */
    fun forGender(gender: NounGender): String = when (gender) {
        NounGender.M -> word
        NounGender.F, NounGender.N -> feminine
    }

    /** Returns the masculine form of this adjective. */
    fun forMasculine(): String = word

    /** Returns the neuter/feminine form of this adjective (same in Romanian). */
    fun forNeuterOrFeminine(): String = feminine

    companion object {
        fun computeFeminine(word: String): String {
            require(!word.isBlank()) { "Adjective feminine derivation requires non-blank input" }
            return when {
                word == "negru" -> "neagră"
                word == "roșu" -> "roșie"
                word == "sec" -> "seacă"
                word == "des" -> "deasă"
                word == "drept" -> "dreaptă"
                word == "întreg" -> "întreagă"
                word == "deșert" -> "deșeartă"
                word == "mort" -> "moartă"
                word.endsWith("esc") -> word.dropLast(2) + "ască"
                word.endsWith("eț") -> word.dropLast(1) + "ață"
                word.endsWith("tor") -> word.dropLast(2) + "oare"
                word.endsWith("șor") -> word.dropLast(2) + "oară"
                word.endsWith("ior") -> word.dropLast(2) + "oară"
                word.endsWith("os") -> word.dropLast(1) + "asă"
                word.endsWith("iu") -> word.dropLast(1) + "e"
                word.endsWith("ci") -> word.dropLast(1) + "e"
                word.endsWith("ru") -> word.dropLast(1) + "ă"
                word.endsWith("țel") || word.endsWith("șel") || word.endsWith("rel") ->
                    word.dropLast(2) + "ică"
                word.endsWith("e") || word.endsWith("o") || word.endsWith("i") -> word
                else -> word + "ă"
            }
        }
    }
}
